File: 109. Convert Sorted List to Binary Search Tree/109. Convert Sorted List to Binary Search Tree.js
// Definition for singly-linked list.
function ListNode(val, next = null) {
    this.val = val;
    this.next = next;
}

// Definition for a binary tree node.
function TreeNode(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
}

function sortedListToBST(head) {
    // Convert linked list to array
    const values = [];
    let current = head;
    while (current !== null) {
        values.push(current.val);
        current = current.next;
    }

    // Helper function to convert sorted array to BST
    function convertArrayToBST(left, right) {
        if (left > right) return null;

        const mid = Math.floor((left + right) / 2);
        const root = new TreeNode(values[mid]);

        root.left = convertArrayToBST(left, mid - 1);
        root.right = convertArrayToBST(mid + 1, right);

        return root;
    }

    return convertArrayToBST(0, values.length - 1);
}

File: 109. Convert Sorted List to Binary Search Tree/109. Convert Sorted List to Binary Search Tree.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn sorted_list_to_bst(head: Option<Box<ListNode>>) -> Option<Rc<RefCell<TreeNode>>> {
        fn to_vec(mut head: Option<Box<ListNode>>) -> Vec<i32> {
            let mut values = vec![];
            while let Some(node) = head {
                values.push(node.val);
                head = node.next;
            }
            values
        }
        
        fn build_bst(values: &[i32]) -> Option<Rc<RefCell<TreeNode>>> {
            if values.is_empty() {
                return None;
            }
            
            let mid = values.len() / 2;
            let mut root = TreeNode::new(values[mid]);
            
            root.left = build_bst(&values[..mid]);
            root.right = build_bst(&values[mid + 1..]);
            
            Some(Rc::new(RefCell::new(root)))
        }
        
        build_bst(&to_vec(head))
    }
}